### ISSUE TITLE
`Cannot call method 'sink' on a null object` when using `loop` in sink context

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#2069]
Golf:
```perl6
sub foo {
    loop (my int $i = 0; $i <= 2; $i++) {
        note ‘hello’
    }
}

for 42 {
    foo
}
```

This was dug out of toaster. The issue was in [Native::Packing](https://github.com/p6-pdf/Native-Packing-p6) module and it required this “fix”: https://github.com/p6-pdf/Native-Packing-p6/commit/23dcfc761d0c0c0b0032d7cb8dce1e32ce387794. I think it's a rakudo bug and we should fix it in rakudo instead.

Bisected: (2018-07-09) https://github.com/rakudo/rakudo/commit/6d271667c9ac2c787e7e20b6170046ccfbd7d7e9

Output on all releases: https://gist.github.com/2f1cbd34b2991d15c7dcda3a98c1e623

Note that the output was different in the past, and the fix was in (2016-08-05) https://github.com/rakudo/rakudo/commit/589061eac14f2847e2c4b401d2ff2eb30c62675e.

**Comment by AlexDaniel** [comment]
Another module that is failing is [Slang::SQL](https://github.com/tony-o/perl6-slang-sql) and bisectable points to the same commit (2018-07-09) https://github.com/rakudo/rakudo/commit/6d271667c9ac2c787e7e20b6170046ccfbd7d7e9. There's no golf for that module, but when the OP issue is fixed I guess one can simply try Slang::SQL's tests to see if it's fixed also. 

When running its tests the output is:
```
===SORRY!===
Cannot call method 'ACCEPTS' on a null object
```

**Comment by jnthn** [comment]
The return value handling changes removed a check for if we somehow ended up with an `nqp::null` being returned, and turning it into a `Mu` instead. This situation should really never happen, and the check was thus papering over other bugs (in the best case, just the one the golf exposes).

In this particular case the mapping to `Mu` also created a discontinuity: a `loop (my int $i = 0; $i <= 2; $i++) { }` as the last statement in a `sub` would end up with the `sub` evaluating to `Mu`, whereas every other kind of loop - and even a `loop` of the form `loop { }` or `loop (; $++ < 3;) { }` - would end up with it evaluating to `Nil`.

I've fixed that in aff96baefd. It's possible that we have some other constructs leaking an `nqp::null`, so testing of the modules in question to make sure they also were due to the `loop` code-gen bug would be very welcome. Even if they're all fine, a test should be added before this issue can be resolved.
